romancal/datamodels/library.py

```diff
@@ -0,0 +1,504 @@
+                        "group_id": group_id,
+                    }
+                )
+
+            # make a fake association
+            self._asn = {
+                # TODO other asn data?
+                "products": [
+                    {
+                        "members": members,
+                    }
+                ],
+            }
+            self._members = self._asn["products"][0]["members"]
+
+            # _asn_dir?
+            # _asn_path?
+
+        elif isinstance(init, self.__class__):
+            # TODO clone/copy?
+            raise NotImplementedError()
+        else:
+            raise NotImplementedError()
+
+        # make sure first model is loaded in memory (as expected by stpipe)
+        if self._asn_n_members == 1:
+            # FIXME stpipe also reaches into _models (instead of _model_store)
+            self._models = [self._load_member(0)]
+
+    def __del__(self):
+        # FIXME when stpipe no longer uses '_models'
+        if hasattr(self, "_models"):
+            self._models[0].close()
+
+    @property
+    def asn(self):
+        # return a "read only" association
+        def _to_read_only(obj):
+            if isinstance(obj, dict):
+                return MappingProxyType(obj)
+            if isinstance(obj, list):
+                return tuple(obj)
+            return obj
+
+        return asdf.treeutil.walk_and_modify(self._asn, _to_read_only)
+
+    # TODO we may want to not expose this as it could go out-of-sync
+    # pretty easily with the actual models.
+    # @property
+    # def members(self):
+    #     return self.asn['products'][0]['members']
+
+    @property
+    def group_names(self):
+        names = set()
+        for member in self._members:
+            names.add(member["group_id"])
+        return names
+
+    @property
+    def group_indices(self):
+        group_dict = {}
+        for i, member in enumerate(self._members):
+            group_id = member["group_id"]
+            if group_id not in group_dict:
+                group_dict[group_id] = []
+            group_dict[group_id].append(i)
+        return group_dict
+
+    def __len__(self):
+        return len(self._members)
+
+    def __getitem__(self, index):
+        if not self._open:
+            raise ClosedLibraryError("ModelLibrary is not open")
+
+        # if model was already borrowed, raise
+        if index in self._ledger:
+            raise BorrowError("Attempt to double-borrow model")
+
+        if index in self._model_store:
+            model = self._model_store[index]
+        else:
+            model = self._load_member(index)
+            if not self._on_disk:
+                # it's ok to keep this in memory since _on_disk is False
+                self._model_store[index] = model
+
+        # track the model is "in use"
+        self._ledger[index] = model
+        return model
+
+    def __setitem__(self, index, model):
```

Comment:
Does the setter method remove the model from memory after saving to the temporary file?  I'm relatively new to looking at this type of code so I'm probably missing it, but I don't see where that happens here or in `self._model_store`

Reply:
For a "in memory" library (when `_on_disk=False`) references to the models are held by the `_model_store`. They won't be candidates for garbage collection until the library is collected. For an "on disk" library the store doesn't hold any references to the models. In the "on disk" mode the models will be removed from memory only after the step removes all references. One of the goals of the "borrow"/"shelve" API is to encourage step developers to be aware of when the step is using a model. It however does not prevent the step from keeping references to the models and preventing them from being freed.